### PR TITLE
Nested struct generation example

### PR DESCRIPTION
**Pull request by lechengfan**
Pretty rough sketch still, but an example of how we'd might want to generate events for those with nested structs. Running `cmd/genabi` gives:

```
type e struct {
	S struct {
		A *big.Int
		B *big.Int
		C struct {
			X *big.Int
			Y *big.Int
		}
	}
}
```

### CODE REVIEW

File: abi/nested/template.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	"text/template"
+
+	"github.com/indexsupply/x/abi"
+)
+
+const ct = `
+{{with .Event}}
+type {{.Name}} struct {
+	{{range .Inputs}}
+		{{geninput .}}
```
Comment (lechengfan) [inline abi/nested/template.go:17]: I didn't end up using `define` since having a custom go function appeared more readable to me

Comment (ryandotsmith) [reply abi/nested/template.go:17]:
```
{{ define "x" }}
{{ if gt (len .Comp) 0 }}
{{ .Name }} struct { 
    {{ range .Comp }}
    {{ template "x" . }}
    {{ end }}
}
{{ else }}
{{ .Name }} {{ .Gtyp }}
{{ end }}
{{ end }}

type {{ template "x" . }}
```

Comment (ryandotsmith) [reply abi/nested/template.go:17]: I kind of like solving it in the template for a couple of reasons:
1) By using another function we spread the template building code into 2 places with effectively two different semantic sets: go/sprintf AND text/emplate. I think it's simpler to either do everything with sprintf or everything with template.
2) The recursive template solution is no less readable then the go function

**Comment by ryandotsmith** [comment]
Closing in favor of #62

**Pull request closed without merging**